src/quic/session.cc

```diff
@@ -0,0 +1,3506 @@
+                         ngtcp2_conn_get_remote_transport_params(connection()));
+}
+
+bool Session::is_in_closing_period() const {
+  return ngtcp2_conn_is_in_closing_period(connection());
+}
+
+bool Session::is_in_draining_period() const {
+  return ngtcp2_conn_is_in_draining_period(connection());
+}
+
+bool Session::is_unable_to_send_packets() const {
+  return NgCallbackScope::InNgCallbackScope(*this) || is_destroyed() ||
+         is_in_draining_period() || (is_server() && is_in_closing_period()) ||
+         !endpoint_;
+}
+
+uint64_t Session::max_data_left() const {
+  return ngtcp2_conn_get_max_data_left(connection());
+}
+
+uint64_t Session::max_local_streams_uni() const {
+  return ngtcp2_conn_get_max_local_streams_uni(connection());
+}
+
+uint64_t Session::max_local_streams_bidi() const {
+  return ngtcp2_conn_get_local_transport_params(connection())
+      ->initial_max_streams_bidi;
+}
+
+quic_version Session::version() const {
+  return ngtcp2_conn_get_negotiated_version(connection());
+}
+
+void Session::ExtendOffset(size_t amount) {
+  DEBUG_ARGS(this, "Extending offset by %" PRIu64 " bytes", amount);
+  ngtcp2_conn_extend_max_offset(connection(), amount);
+}
+
+void Session::ExtendStreamOffset(stream_id id, size_t amount) {
+  DEBUG_ARGS(this,
+             "Extending stream %" PRIi64 " offset by %" PRIu64 " bytes",
+             id,
+             amount);
+  ngtcp2_conn_extend_max_stream_offset(connection(), id, amount);
+}
+
+void Session::UpdateTimer() {
+  // Both uv_hrtime and ngtcp2_conn_get_expiry return nanosecond units.
+  uint64_t expiry = ngtcp2_conn_get_expiry(connection());
+  uint64_t now = uv_hrtime();
+
+  if (expiry <= now) {
+    // The timer has already expired.
+    return OnTimeout();
+  }
+
+  auto timeout = (expiry - now) / NGTCP2_MILLISECONDS;
+
+  DEBUG_ARGS(this, "Setting transmission timer to %" PRIu64, timeout);
+
+  // If timeout is zero here, it means our timer is less than a millisecond
+  // off from expiry. Let's bump the timer to 1.
+  timer_.Update(timeout == 0 ? 1 : timeout);
+}
+
+void Session::OnTimeout() {
+  HandleScope scope(env()->isolate());
+
+  if (is_destroyed()) return;
+
+  DEBUG(this, "Transmission timer has expired");
+  DEBUG_ARGS(
+      this, "In Closing Period? %s", is_in_closing_period() ? "Yes" : "No");
+  DEBUG_ARGS(
+      this, "In Draining Period? %s", is_in_draining_period() ? "Yes" : "No");
+
+  int ret = ngtcp2_conn_handle_expiry(connection(), uv_hrtime());
+  if (NGTCP2_OK(ret) && !is_in_closing_period() && !is_in_draining_period()) {
+    DEBUG(this, "Retransmitting");
+    SendPendingDataScope send_scope(this);
+    return;
+  }
+
+  DEBUG(this, "Closing silently after timeout");
+  last_error_ = QuicError::ForNgtcp2Error(ret);
+  CloseSilently();
+}
+
+void Session::SendConnectionClose() {
+  CHECK(!NgCallbackScope::InNgCallbackScope(*this));
+
+  if (is_destroyed() || is_in_draining_period() || state_->silent_close) return;
```

Comment:
This fix is the same as I found in https://github.com/HalleyAssist/node/commit/f505f7a325746c58d551906ee7f6c19b93ad16b4